apollo-router-core/src/error.rs

```diff
@@ -133,6 +130,88 @@ pub struct Error {
     pub extensions: Object,
 }
 
+impl Error {
+    pub fn from_value(service_name: &str, value: Value) -> Result<Error, graphql::FetchError> {
+        let mut object = match value {
+            Value::Object(o) => o,
+            _ => {
+                return Err(graphql::FetchError::SubrequestMalformedResponse {
+                    service: service_name.to_string(),
+                    reason: "expected a JSON object".to_string(),
+                })
+            }
+        };
+
+        let message = match object
+            .get("message")
+            .and_then(|v| v.as_str())
+            .map(|s| s.to_string())
+        {
+            Some(s) => s,
+            None => {
+                return Err(graphql::FetchError::SubrequestMalformedResponse {
+                    service: service_name.to_string(),
+                    reason: "missing message field".to_string(),
+                })
+            }
+        };
+
+        let path = object
+            .get("path")
+            .and_then(|v| v.as_str())
+            .map(|s| s.into());
+
+        let locations = Vec::new();
+
+        let mut errors = Vec::new();
+        match object.remove("locations") {
+            Some(Value::Array(v)) => {
+                for val in v.into_iter() {
+                    if let Value::Object(o) = val {
+                        if let (Some(line), Some(column)) = (
+                            o.get("line").and_then(|v| v.as_i64()),
+                            o.get("column").and_then(|v| v.as_i64()),
+                        ) {
+                            errors.push(Location {
+                                line: line as i32,
+                                column: column as i32,
+                            });
+                            continue;
+                        }
+                    }
+                    return Err(graphql::FetchError::SubrequestMalformedResponse {
+                        service: service_name.to_string(),
+                        reason: "missing location value".to_string(),
+                    });
+                }
+            }
+            _ => {
+                return Err(graphql::FetchError::SubrequestMalformedResponse {
+                    service: service_name.to_string(),
+                    reason: "missing errors field".to_string(),
+                })
+            }
+        };
+
+        let extensions = match object.remove("extensions") {
+            Some(Value::Object(o)) => o,
+            _ => {
+                return Err(graphql::FetchError::SubrequestMalformedResponse {
+                    service: service_name.to_string(),
+                    reason: "missing extensions field".to_string(),
+                })
+            }
+        };
+
+        Ok(Error {
+            message,
+            locations,
+            path,
+            extensions,
+        })
+    }
+}
```

Comment:
:yarn: so:
- if you are in the core crate, everything gets imported into scope from the crate itself. Doing `use crate::prelude::graphql::*` === `use crate::*`. In graphql core, there is no reason to distinguish between graphql and non-graphql stuff (we don't have `http::Request` here, we only have `graphql::Request`, so the `graphql::` prefix is useless)
- if you are not in the core crate, you want to make it explicit that something is "graphql stuff" so you `use crate::prelude::*` and prefix all graphql related stuff with `graphql::` (in router we do have `http::Request` and `graphql::Request`, the `graphql::` prefix is useful for that reason)

```suggestion
    pub fn from_value(service_name: &str, value: Value) -> Result<Error, FetchError> {
        let mut object = match value {
            Value::Object(o) => o,
            _ => {
                return Err(FetchError::SubrequestMalformedResponse {
                    service: service_name.to_string(),
                    reason: "expected a JSON object".to_string(),
                })
            }
        };

        let message = match object
            .get("message")
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
        {
            Some(s) => s,
            None => {
                return Err(FetchError::SubrequestMalformedResponse {
                    service: service_name.to_string(),
                    reason: "missing message field".to_string(),
                })
            }
        };

        let path = object
            .get("path")
            .and_then(|v| v.as_str())
            .map(|s| s.into());

        let locations = Vec::new();

        let mut errors = Vec::new();
        match object.remove("locations") {
            Some(Value::Array(v)) => {
                for val in v.into_iter() {
                    if let Value::Object(o) = val {
                        if let (Some(line), Some(column)) = (
                            o.get("line").and_then(|v| v.as_i64()),
                            o.get("column").and_then(|v| v.as_i64()),
                        ) {
                            errors.push(Location {
                                line: line as i32,
                                column: column as i32,
                            });
                            continue;
                        }
                    }
                    return Err(FetchError::SubrequestMalformedResponse {
                        service: service_name.to_string(),
                        reason: "missing location value".to_string(),
                    });
                }
            }
            _ => {
                return Err(FetchError::SubrequestMalformedResponse {
                    service: service_name.to_string(),
                    reason: "missing errors field".to_string(),
                })
            }
        };

        let extensions = match object.remove("extensions") {
            Some(Value::Object(o)) => o,
            _ => {
                return Err(FetchError::SubrequestMalformedResponse {
                    service: service_name.to_string(),
                    reason: "missing extensions field".to_string(),
                })
            }
        };

        Ok(Error {
            message,
            locations,
            path,
            extensions,
        })
    }
}
```

Reply:
alright, the practice around importing or not `graphql::` was not obvious to me, thanks for explaining.
I expect to remove that function in the end: I wrote it to avoid adding an entire `Deserializer` implementation for `Value`: https://github.com/serde-rs/json/blob/master/src/value/de.rs#L700

Reply:
fixed in 4878014